enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,940 @@
+superseded-by:
+  - None
+---
+
+# MicroShift updateability in ostree based systems
+
+## Summary
+
+This enhancement focuses on high level overview of updating
+MicroShift running on ostree based systems such as RHEL 4 Edge.
+Enhancement covers backup and restore of MicroShift data,
+version migration (upgrade and downgrade) of MicroShift and
+its consequences (migration of data between schema versions),
+and interactions with GreenBoot and operating system.
+
+## Motivation
+
+MicroShift team is working towards a general availability (GA) release.
+As GA product, it is expected it can be updated to
+apply security patches and bug fixes, and leverage features
+in newer version while keeping and using existing data.
+
+MicroShift is intended to be a part of Red Hat Device Edge (RHDE) which is based
+on RHEL For Edge (R4E) which is immutable Linux distribution by leveraging
+[ostree](https://ostreedev.github.io/ostree/) technology.
+It allows changing root filesystem for next boot by staging new commits or
+rolling back to previous one.
+ostree is commonly paired with [greenboot](https://github.com/fedora-iot/greenboot)
+which provides automated health assessment of the system and trigger a rollback
+if rebooting the device doesn't result in device becoming healthy.
+
+Even though, OpenShift does not support downgrade or rollback, MicroShift must
+support it in some form to fit into RHDE.
+Rollback (going back to older deployment) will be supported only if MicroShift
+ran on that deployment and data compatible with that deployment was backed up.
+Downgrade (migrating to older version of MicroShift) will not be supported.
+
+In order to integrate into RHDE, MicroShift needs to be augmented with
+functionality to back up and restore its data together with ostree deployments,
+and refuse or perform data migration to newer storage schema version.
+Integration with greenboot will allow system's health to depend on state of the
+MicroShift and will provide necessary information to manage backups.
+
+### User Stories
+
+* As a MicroShift administrator, I want to safely update MicroShift
+  so that I can get bug fixes, new features, and security patches.
+* As a MicroShift administrator, I want my system to roll back
+  to a previous good configuration when greenboot fails.
+
+### Goals
+
+Goal of the enhancement is to describe implementation roadmap for
+integrating MicroShift with ostree and greenboot in order to provide
+functionality to:
+- Safely update MicroShift version (by backing up the data and
+  restoring it in case of rollback)
+- Migrating internal data (like Kubernetes storage or etcd schema) to
+  newer version
+- Block data migration if MicroShift version skew is unsupported
+
+Design aims to implement following principles:
+- Keep it simple, optimize later
+- MicroShift does not own the OS or host
+- MicroShift and all its components are versioned, upgraded, and rolled back together
```

Comment:
Yes it is.
Until earlier this year etcd was included in the `microshift` binary, currently it's a separate binary (although it's not upstream etcd binary but "etcd server imported and started by layer that reads MicroShift config"), but it's shipped in the same RPM. This also means that specific release of microshift includes specific version of etcd.
From that point of view it's included in "MicroShift's components". I'll mention it explicitly

Reply:
i was referring to the etcd data, not the etcd binary.  the etcd data is not part of "microshift"(not part of the distribution).

so i was expecting to see an explicit reference to the backup/restore of the etcd data since that is a key part of this solution.

Reply:
I see what you meant now.
Currently we stored etcd data in `/var/lib/microshift/etcd` and everything related to backup,restore, and migration is performed on `/var/lib/microshift` directory generally. There's no special treatment for etcd's data dir.

This enhancement left out specifics like paths on purpose, but current plan is to basically run `cp --reflinks=auto --preserve --recursive --verbose /var/lib/microshift /var/lib/microshift-backups/ostree-deploy-id/` to backup and similar to restore.
As for migration of etcd db schema - I mentioned it somewhere, but (looking back at v2 and v3) this shouldn't happen until etcd v4 comes out

Reply:
is there any chance a customer might want to configure the path of the backups?  such as to use a network attached device or just a different disk/partition that might have more capacity?

Reply:
We could support putting the data somewhere other than the same filesystem in the future. True host backups, though, are out of scope for this design, and are being handled separately. They may end up using different tools, triggers, etc. We would not, I think, support automated backup and restore as part of MicroShift, but would expect someone doing that to provide their own automation.